docker_cli/commands.py

```diff
@@ -66,6 +67,61 @@ def get_archive(
     return Image.build(tag=img_tag, dockerfile_string=dockerfile, no_cache=True)
 
 
+def insert(tag: str, base: str, directory: str | os.PathLike[str]):
+    """
+    Builds a Docker image with the contents of the given directory copied onto it.
+
+    The directory path on the image has the same name as the topmost directory
+    of the given path. e.g. giving path "/tmp/dir/subdir" will result in the contents of
+    this path being saved in "/subdir" on the generated image.
+
+    This Dockerfile also places the working directory of the image inside of the copied
+    directory.
+
+    Parameters
+    ----------
+    tag : str
+        The image tag.
+    base : str
+        The base image tag.
+    directory : path-like
+        The directory to be copied.
+
+    Returns
+    -------
+    Image
+        The generated image.
+    """
+
+    prefix = universal_tag_prefix()
+    img_tag = tag if tag.startswith(prefix) else f"{prefix}-{tag}"
+
+    dir_str = str(directory)
+
+    # The absolute path of the given directory will be the build context.
+    # This is necessary because otherwise docker may be unable to find the directory if
+    # the build context is at the current working directory.
+    path_absolute = os.path.abspath(str(dir_str))
+    # Additionally, the top directory of the given path will be the name of the
+    # directory in the image.
+    if os.path.isdir(dir_str):
+        target_dir = os.path.basename(path_absolute)
+    else:
+        target_dir = os.path.basename(os.path.dirname(path_absolute))
+
+    # Generate the dockerfile.
+    dockerfile: str = insert_dir_dockerfile(
+        base=base,
+        target_dir=target_dir,
+        source_dir=".",
+    )
+
+    # Build the image with the context at the absolute path of the given path.
+    return Image.build(
+        tag=img_tag, context=path_absolute, dockerfile_string=dockerfile, no_cache=True
```

Comment:
>`no_cache=True`

Why are we disabling the cache here?

Reply:
We had a discussion some time ago in which you mentioned that the build steps would all be run with the cache disabled. I interpreted that to include this step - at that time, did you mean that the cache would only be disabled for the steps following this one?

Reply:
Oh yeah, thanks for the reminder, you're right -- I do remember thinking that we might need to disable the cache for these build steps.

I think ultimately the desired behavior is this: if a user makes an update to the source code in their local repository on the host and then re-runs the build steps via wigwam, then Docker should correctly recognize that the code has changed and needs to be recompiled (rather than using a previously cached image, even though the Dockerfile didn't change).

But, in hindsight, Docker already handles that case correctly even if we don't manually disable the cache. As long as the user re-runs the ~`insert`~ `copydir` command after changing the source code, Docker will recognize that the contents of the directory have changed and will invalidate the cache.

So I can't quite remember exactly why I thought that we wanted to disable the cache. Maybe I wanted to ensure that the CI system would always rebuild the code on each PR update rather than using a cached image? But that would be the case anyway unless the PR was empty somehow (with no source code changes). Maybe I just didn't think it through properly or else had a different mental model for how this would work than what we ended up settling on :shrug:

I guess I would suggest removing the `no_cache=True` here unless you have a better memory than me about why we were thinking about disabling it. I can't think of any benefit to disabling it right now.